pkg/cosign/tuf/client.go

```diff
@@ -475,25 +449,63 @@ func (t *TUF) updateMetadataAndDownloadTargets() error {
 }
 
 type targetDestination struct {
-	buf bytes.Buffer
+	buf *bytes.Buffer
 }
 
 func (t *targetDestination) Write(b []byte) (int, error) {
 	return t.buf.Write(b)
 }
 
 func (t *targetDestination) Delete() error {
-	t.buf = bytes.Buffer{}
+	t.buf = &bytes.Buffer{}
 	return nil
 }
 
-func downloadRemoteTarget(name string, c *client.Client, w io.Writer) error {
-	dest := targetDestination{}
-	if err := c.Download(name, &dest); err != nil {
-		return fmt.Errorf("downloading target: %w", err)
+func maybeDownloadRemoteTarget(name string, meta data.TargetFileMeta, t *TUF) error {
+	// If we already have the target locally, don't bother downloading from remote storage.
+	if cachedTarget, err := t.targets.Get(name); err == nil {
+		// If the target we have stored matches the meta, use that.
+		if isValidTarget(cachedTarget, meta) {
+			return nil
+		}
+	}
+
+	// Check if we already have the target in the embedded store.
+	w := bytes.Buffer{}
+	rd, ok := t.embedded.(fs.ReadFileFS)
+	if !ok {
+		return errors.New("fs.ReadFileFS unimplemented for embedded repo")
+	}
+	b, err := rd.ReadFile(path.Join("repository", "targets", name))
+
+	if err == nil {
+		// Unfortunately go:embed appears to somehow replace our line endings on windows, we need to switch them back.
+		// It should theoretically be safe to do this everywhere - but the files only seem to get mutated on Windows so
+		// let's only change them back there.
+		if runtime.GOOS == "windows" {
+			b = bytes.ReplaceAll(b, []byte("\r\n"), []byte("\n"))
+		}
+
+		if isValidTarget(b, meta) {
+			if _, err := io.Copy(&w, bytes.NewReader(b)); err != nil {
+				return fmt.Errorf("using embedded target: %w", err)
+			}
+		}
 	}
-	_, err := io.Copy(w, &dest.buf)
-	return err
+
+	// Nope -- no local matching target, go download it.
+	if w.Len() == 0 {
+		dest := targetDestination{buf: &w}
+		if err := t.client.Download(name, &dest); err != nil {
+			return fmt.Errorf("downloading target: %w", err)
+		}
```

Comment:
Do we need to also check `isValidTarget`, or will `Download` handle checking the target hash? (If Download doesn't, it should!)

Reply:
Correct, it does check hash handling!
https://github.com/theupdateframework/go-tuf/blob/3b26aedfe985198bc88a9dda7525938c575ca046/client/client.go#L928-L934

I also put hash checking on the `Get`/retrieval side, just for robustness: https://github.com/sigstore/cosign/blob/e7bcb6961178788ceb5a2e63e33388a188b574c9/pkg/cosign/tuf/client.go#L343-L354